examples/imitation_learning_openbot/generate_episodes.py

```diff
@@ -42,71 +43,73 @@
 
-            # reset the simulation
-            env_reset_info = {}
-            _ = env.reset(reset_info=env_reset_info)
-            assert "success" in env_reset_info
-            
-            # if it took too long to reset the simulation, then continue
-            if not env_reset_info["success"]:
-                spear.log("Call to env.reset(...) was not successful. Simulation took too long to return to a ready state. Retrying...")
-                continue
-                
-            # get random start-goal pairs
-            _, _, _, env_step_info = env.step(action={"apply_voltage": np.array([0.0, 0.0], dtype=np.float32)})
-
-            positions = env_step_info["agent_step_info"]["trajectory_data"] 
-
-            # store poses in a csv file
-            df = pd.DataFrame({"scene_id"       : [scene_id],
-                               "init_pos_x_cms" : positions[0,0],
-                               "init_pos_y_cms" : positions[0,1],
-                               "init_pos_z_cms" : positions[0,2],
-                               "goal_pos_x_cms" : positions[-1,0],
-                               "goal_pos_y_cms" : positions[-1,1],
-                               "goal_pos_z_cms" : positions[-1,2]})
-            df.to_csv(args.episodes_file, mode="w" if i==0 and j==0 else "a", index=False, header=i==0 and j==0)
+        # choose only the top args.num_episodes_per_scene trajectories
+        start_points_sorted = candidate_points[sorted_indicies]        
+        top_start_points = start_points_sorted[-args.num_episodes_per_scene:]
         
-            plt.plot(positions[0,0], positions[0,1], "^", markersize=12.0, label="Start", color="tab:blue", alpha=0.3)
-            plt.plot(positions[-1,0], positions[-1,1], "^", markersize=12.0, label="Goal", color="tab:orange", alpha=0.3)
-            plt.plot(positions[:,0], positions[:,1], "-o", markersize=8.0, label="Desired trajectory", color="tab:green", alpha=0.3)
-
-            j = j+1
+        end_points_sorted = reachable_points[sorted_indicies]        
+        top_end_points = end_points_sorted[-args.num_episodes_per_scene:]
+
+        # concat arrays for easier pd dataframe creation
+        merged_array = np.hstack((top_start_points, top_end_points))
+
+        # store start and end location of the trajectories
+        df_ = pd.DataFrame(
+            columns=df_columns,
+            data={"scene_id"         : [scene_id] * merged_array.shape[0],
+                  "start_location_x" : merged_array[:,0],
+                  "start_location_y" : merged_array[:,1],
+                  "start_location_z" : merged_array[:,2],
+                  "goal_location_x"  : merged_array[:,3],
+                  "goal_location_y"  : merged_array[:,4],
+                  "goal_location_z"  : merged_array[:,5]})
+        
+        df = pd.concat([df, df_])
 
+        plt.plot(merged_array[:,0], merged_array[:,1], "^", markersize=12.0, label="Start", color="tab:blue", alpha=0.3)
```

Comment:
move this plotting code into `utils.py` for consistency with the other plotting code

Reply:
Updated this.